### PR TITLE
HDDS-10000. Improve LeakDetector

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

 * Let `LeakTracker` implement `UncheckedAutoCloseable` and change methods that return it to use the interface.  Tracked objects should not call other methods exposed due to `LeakTracker extends WeakReference`.
 * Distinguish detectors with the same name by appending a unique integer.
 * Split two test cases in `TestLeakDetector`.

https://issues.apache.org/jira/browse/HDDS-10000

## How was this patch tested?

CI:
https://github.com/adoroszlai/ozone/actions/runs/7401596618

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/LeakDetector.java [hunk 1]
```diff
@@ -18,13 +18,15 @@
  */
 package org.apache.hadoop.hdds.utils;
 
+import org.apache.ratis.util.UncheckedAutoCloseable;
```
Comment (duongkame) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/LeakDetector.java:21]: I don't love that Ozone depends on ratis for such a basic need. I guess it's ok for this PR as it's a general issue in Ozone.

**Review by duongkame — APPROVED** [review]
Thanks for the patch @adoroszlai , LGTM.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/LeakDetector.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/LeakDetector.java:21]: For now I think of Ratis as a library that provides Raft-implementation as well as useful utilities used by both Ratis and Ozone.  The problem is that we don't have a top-level "utilities" Apache (or Apache-licensed other) project.

**Comment by adoroszlai** [comment]
Thanks @duongkame for the review.

**Pull request merged**